PORT-10848 Guide - Working Agreements (TF) (#1659)
port-labs/port-docs: docs/guides/all/working_agreements_and_measuring_pr_standards.md

---
sidebar_position: 1
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

> # Implement working agreements and measure pull request standards 

Reviewer: I wonder if we should name it just "Measure Pull Requests Standards"

Author: makes sense

This guide is aimed at helping engineering teams implement working agreements and measure pull request (PR) standards using Port.
We will implement working agreements using Port's scorecards and measure PR standards using aggregation properties.

> <img src='/img/guides/scorecardOverview.png' border='1px' />

Reviewer: this was broken to me

- [**PR Requires Reviewers**](#pr-requires-reviewers): Confirms that at least one reviewer is assigned to the PR.
- [**PR Is Linked to a Milestone**](#pr-is-linked-to-a-milestone): Ensures the PR is associated with a milestone.
- [**PR Changed X Files or Less**](#pr-changed-x-files-or-less): Validates that the number of changed files is within acceptable limits.
- [**PR Has Been Open for X Days**](#pr-has-been-open-for-x-days): Monitors how long a PR has been open.
- [**PR Batch Size Calculation**](#pr-batch-size-calculation): Calculates the batch size of the PR.

These checks are implemented using Port's [scorecards](/#scorecards) on the `pull request` blueprint.

## Implementation 
This section will guide you through implementing the working agreements and PR checks in your Port environment.
Follow the steps below to  implement this :
1. Add the properties to the `pull request` blueprint.
   - Go to the [Builder](https://app.getport.io/settings/data-model) in your Port portal.
   -  Select the `pull request` blueprint.
   -  Click on the `{...}` button in the top right corner, and choose "Edit JSON".

>    - Add the properties and mapping configurations as described below.

Reviewer: Here I felt like I'm supposed to add properties based on the image. Maybe the image can be at the end of this section (after "Save the changes") with explanation of "after implementing all the checks in the next section ,it will look something like this"

Author: okay

     <img src='/img/guides/prBPPropertyAdd.png'  width='80%' />
   - Save the changes.

2. Add the **mapping configuration** to the data source.
   - Go to the [Data Sources](https://app.getport.io/settings/data-sources) in your Port portal.
   - Select the data source connected to your GitHub repository
   - Add the mapping configurations as described below.
     <img src='/img/guides/addMappingConfigDS.png'  />
   - Save the changes.

3. Add the **scorecard** definitions to the `pull request` blueprint.
   - Go to the [Builder](https://app.getport.io/settings/data-model) in your Port portal.
   - Select the `pull request` blueprint.
   - Click on  the **Scorecard**
   - Click on the `+ New Scorecard` button.

>    <img src='/img/guides/scorecardOnPRAdd.png' />

Reviewer: this was broken for me

5. Select `sum` for the **Function**.

   <img src="/img/guides/totalPrsMerged.png" width="50%"/>

6. Click `Save`.

</details>

<details>
<summary><b> Mean Time to Merge (Days) </b></summary>

1. Click `+ Widget` and select **Number Chart**.
2. Title: `Mean Time to Merge (Days)`, (add the `Merge` icon).
3. Select `Display single property` and choose **Service** the **Entity**.
4. Choose `Averge Time to Merge` as the **Property**.

   <img src="/img/guides/meanTimeToMergePr.png" width="50%"/>

5. Click `Save`.

</details>

<details>
<summary><b> Total Weekly  Pr commits </b></summary>

1. Click `+ Widget` and select **Number Chart**.
2. Title: `Total Weekly  Pr commits`, (add the `GitPullRequest` icon).
3. Select `Aggregate by property` and choose **Service** as the **Blueprint**.
4. Choose `Averge Commits per PRs` as the **Property**.
5. Select `average` for the **Function** and choose `week` for **Average of**.
6. Select `Created At` for **Measure time by**.
7. Select `custom` as the **Unit** and input `Weekly` as the **Custom unit**.

   <img src="/img/guides/totalWeeklyPrCommits.png" width="50%"/>

8. Click `Save`.

</details>

<details>

> <summary><b>Service Scorecard Performance by Team</b></summary>

Reviewer: This should be the most powerful example. If we can build a screenshot of it (will be in the demo next week) and even add at the beginning of this guide and explaining something like "by the end of this guide you'll be able to track teams performance based on pull requests metrics as shown in this screenshot"

---
sidebar_position: 1
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

# Measuring pull request standards
This guide is aimed at helping engineering teams implement working agreements and measure pull request (PR) standards using Port.
We will implement working agreements using Port's scorecards and measure PR standards using aggregation properties.
by the end of this guide you'll be able to track teams performance based on pull requests metrics as shown in this screenshot

> <img src='/img/guides/teamPRMetricTable.png' border='1px' />

Reviewer: This is still borken for me.Can you check?

Author: It's not broken in the preview too

Author: https://pr-1659.d2ngvl90zqbob8.amplifyapp.com/guides/all/working_agreements_and_measuring_pr_standards

## Overview

Effective collaboration and clear expectations are crucial for high-performing engineering teams. **Working agreements** establish shared processes and standards, enhancing teamwork.
Measuring **Pull request (PR)** standards is essential for assessing code quality, review processes, and team efficiency. 
By integrating working agreements with measurable PR metrics, teams can monitor adherence to best practices and continuously improve workflows. 
We will discuss how to implement working agreements and measure PR standards using Port.

:::info Metrics
Metrics are essential for assessing how well teams adhere to their working agreements. 
They enable teams to track compliance, identify bottlenecks, and drive continuous improvement. 

> For detailed insights into key metrics like `deployment frequency`, `lead time for changes`, and `change failure rate`,

Reviewer 2: Same here, the sentence "For detailed...." should be in a new line.
To add a new line, add 2 spaces after the last word. For example:

this is the first line  
This is the second line, note the 2 spaces at the end of the first one

please refer to our [DORA Metrics guide](/guides/all/setup-dora-metrics).
:::

## Prerequisites
- Complete the [Port onboarding process](https://docs.getport.io/quickstart).
- Access to a GitHub repository that is connected to Port via the onboarding process.

## Working agreements

> The following working agreements and PR checks have been implemented in our [demo environment](https://demo.getport.io/scorecard_overview) all the way down for similar examples:

Reviewer: all the way down for similar examples

Not sure I understand the context

Author: yeah yeah let me change that. it was in relation to the old picture.

- [**PR Description Cannot be Empty**](#pr-description-cannot-be-empty): Ensures that every PR has a description.
- [**PR Has Linked Issue**](#pr-has-linked-issue): Verifies that each PR is linked to an issue.
- [**PR Has No Unchecked Checkboxes**](#pr-has-no-unchecked-checkboxes): Checks that there are no unchecked items in the PR description.
- [**PR Requires Reviewers**](#pr-requires-reviewers): Confirms that at least one reviewer is assigned to the PR.
- [**PR Is Linked to a Milestone**](#pr-is-linked-to-a-milestone): Ensures the PR is associated with a milestone.
- [**PR Changed X Files or Less**](#pr-changed-x-files-or-less): Validates that the number of changed files is within acceptable limits.
- [**PR Has Been Open for X Days**](#pr-has-been-open-for-x-days): Monitors how long a PR has been open.
- [**PR Batch Size Calculation**](#pr-batch-size-calculation): Calculates the batch size of the PR.

> These checks are implemented using Port's [scorecards](/#scorecards).

Reviewer 2: fix the link to the correct scorecards page

Author: okay

## Implementation 
This section will guide you through implementing the working agreements and PR checks in your Port environment.

> Follow the steps below to  implement this :

Reviewer 2: ```suggestion
Follow the steps below:
```

1. Add the properties to the `pull request` blueprint.
   - Go to the [Builder](https://app.getport.io/settings/data-model) in your Port portal.
   -  Select the `pull request` blueprint.

>    -  Click on the `{...}` button in the top right corner, and choose "Edit JSON".

Reviewer 2: ```suggestion
   -  Click on the `...` button in the top right corner, and choose "Edit JSON".
```

   - Add the properties and mapping configurations as described below.

>    - Save the changes.

Reviewer 2: note that there is no space between the list and the admonition. This looks bad, use `<br />` where needed to add a gap between them.
Fix this for other places in the guide as well.

>    :::tip Aftermath

Reviewer 2: Aftermath is not a good name for this, please change it.
Also, "implementing all the checks" is not clear. Make it clearer that this is the expected JSON definition.

     After implementing all the checks in the next section, it will look something like this
     <img src='/img/guides/prBPPropertyAdd.png'  width='80%' border="1px" />
   :::

2. Add the **mapping configuration** to the data source.
   - Go to the [Data Sources](https://app.getport.io/settings/data-sources) in your Port portal.
   - Select the data source connected to your GitHub repository
   - Add the mapping configurations as described below.
     <img src='/img/guides/addMappingConfigDS.png' border="1px"  />
   - Save the changes.

3. Add the **scorecard** definitions to the `pull request` blueprint.
   - Go to the [Builder](https://app.getport.io/settings/data-model) in your Port portal.
   - Select the `pull request` blueprint.
   - Click on  the **Scorecard**
   - Click on the `+ New Scorecard` button.

>    <img src='/img/guides/scorecardOnPRAdd.png' border="1px" />

Reviewer: This image is broken for me. I'm checking in the guide link in the PR

Author: I made sure everything was pushed. let me check again

Author: it's not like that in the preview link.

   - Paste the scorecard definitions as described below.
   - Save the changes.

### PR Description Cannot be Empty

<h4> Scorecard Definition </h4>

<details>

>   <summary>Click to view the scorecard definition</summary>

Reviewer 2: change all of these to be in our regular format:
- should be bold, using <b></b>
- should say "Scorecard definition (click to expand)"

  <summary>Click to view the aggregation property</summary>

```json
"averageCommitsPerPr": {
  "title": "Average Commits per PR",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "averageOf": "week",
    "func": "average",
    "calculationBy": "property",
    "property": "commits",
    "measureTimeBy": "$createdAt"
  }
}
```

</details>

</TabItem>

<TabItem value="averageLocChanged">

Add the following aggregation property to the **service** and **organization** blueprints:

<details>
  <summary>Click to view the aggregation property</summary>

```json
"averagePrLinesOfCode": {
  "title": "Average PR Lines of Code",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "func": "average",
    "averageOf": "week",
    "calculationBy": "property",
    "property": "totalLocChanged"
  }
}
```

</details>

</TabItem>

</Tabs>

By implementing these aggregation properties, you can effectively measure and monitor PR standards at both the service and organization levels.

## Visualization
By leveraging Port's Dashboards, you can create custom dashboards to track the pr metrics and monitor your team's performance over time.

> <img src='/img/guides/scorecardsOverview.png' border='1px' />

Reviewer: This visualization shows properties, not scorecards. These properties are interesting in widgets context but here I thought we will create scorecard for the service blueprint and show the nice aggregation with colors as in the Demo (if it's not like that we will update the demo)

Author: @Matanga1-2 can you check it now

Author: @Matanga1-2 Can you check it now ? please

   - Go to the [Builder](https://app.getport.io/settings/data-model) in your Port portal.
   -  Select the `pull request` blueprint.
   -  Click on the `...` button in the top right corner, and choose "Edit JSON".
   - Add the properties and mapping configurations as described below.
   - Save the changes.
   
   <br/>
   
   :::tip Expected JSON  
     This is the expected JSON definition after adding the properties.
     <img src='/img/guides/prBPPropertyAdd.png'  width='80%' border="1px" />
   :::
<br/>
2. Add the **mapping configuration** to the data source.
   - Go to the [Data Sources](https://app.getport.io/settings/data-sources) in your Port portal.
   - Select the data source connected to your GitHub repository
   - Add the mapping configurations as described below.
     <img src='/img/guides/addMappingConfigDS.png' border="1px"  />
   - Save the changes.

3. Add the **scorecard** definitions to the `pull request` blueprint.
   - Go to the [Builder](https://app.getport.io/settings/data-model) in your Port portal.
   - Select the `pull request` blueprint.
   - Click on  the **Scorecard**
   - Click on the `+ New Scorecard` button.

>    <img src='/img/guides/scorecardOnPRAdd.png' border="1px" />

Reviewer 2: ```suggestion
     <img src='/img/guides/scorecardOnPRAdd.png' border="1px" />
```

>      <img src='/img/guides/scorecardOnPRAdd.png' border="1px" />

Reviewer 2: ```suggestion
   <img src='/img/guides/scorecardOnPRAdd.png' border="1px" />
```

"issueUrl": {
  "title": "Issue URL",
  "type": "string",
  "format": "url"
}
```

</details>

<h4> Mapping Configuration </h4>
Map the issue URL from your data source to the `issueurl` property:

<details>
  <summary><b>Mapping config (click to expand)</b></summary>

```yaml showLineNumbers
  - kind: pull-request
    selector:
      query: 'true'
    port:
      entity:
        mappings:
          identifier: .head.repo.name + (.id|tostring)
          title: .title
          blueprint: '"githubPullRequest"'
          properties:
            #other properties
            issueUrl: .issue_url
```

</details>

### PR Has No Unchecked Checkboxes

<h4> Scorecard Definition </h4>

<details>
  <summary><b>Scorecard definition (click to expand)</b></summary>

```json
{
  "identifier": "pr_no_unchecked_chk",
  "title": "PR Has No Unchecked Checkboxes",
  "description": "Ensures that there are no unchecked checkboxes in the PR description.",
  "level": "Silver",
  "query": {
    "combinator": "and",
    "conditions": [
      {
        "operator": "doesNotContains",
        "property": "prDescription",
        "value": "- [ ]"
      }
    ]
  }
}
```

</details>

:::tip Property addition

> If you haven't added the `prDescription` property, and it's relative mapping config

Reviewer 2: ```suggestion
If you haven't added the `prDescription` property and its relative mapping config,
```

---
sidebar_position: 1
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

# Implement working agreements and measure pull request standards 

## Overview

Creating high-performing engineering teams requires not just technical expertise but also effective collaboration and clear expectations.

> **working agreements** help teams establish shared understandings of processes and standards, enhancing collaboration, promoting accountability, 

Reviewer: ```suggestion
**working agreements** help teams establish shared understandings of processes and standards while enhancing collaboration.
```

Reviewer: I think we can make it more concise

> facilitating onboarding, and improving efficiency.**pull requests (PRs)** are a critical part of the workflow. Measuring PR standards is essential for assessing code quality,

Reviewer: ```suggestion
**Pull requests (PRs)** are a critical part of the workflow. Measuring PR standards is essential for assessing code quality,
```

I think measuring PR standards is essential to improve efficiency metrics and make sure teams are working within defined standards.

review processes, and team efficiency. By integrating **working agreements** and **measurable PR metrics**, 

> teams can monitor adherence to best practices and continuously improve their workflows.

Reviewer: We can make this entire paragraph more concise

Author: okay

We will discuss how to implement working agreements and measure PR standards using Port.

:::info Metrics
Metrics are essential for assessing how well teams adhere to their working agreements. 
They enable teams to track compliance, identify bottlenecks, and drive continuous improvement. 
For detailed insights into key metrics like `deployment frequency`, `lead time for changes`, and `change failure rate`,
please refer to our [DORA Metrics guide](/guides/all/setup-dora-metrics).
:::

## Prerequisites
- Complete the [Port onboarding process](https://docs.getport.io/quickstart).
- Access to a GitHub repository that is connected to Port via the onboarding process.

## Working agreements

> The following working agreements and pr checks have been implemented in our [demo environment](https://demo.getport.io/settings/data-model) on the `pull request` blueprint:

Reviewer: PR? (not pr)

- [**PR Description Cannot be Empty**](#pr-description-cannot-be-empty): Ensures that every PR has a description.
- [**PR Has Linked Issue**](#pr-has-linked-issue): Verifies that each PR is linked to an issue.
- [**PR Has No Unchecked Checkboxes**](#pr-has-no-unchecked-checkboxes): Checks that there are no unchecked items in the PR description.
- [**PR Requires Reviewers**](#pr-requires-reviewers): Confirms that at least one reviewer is assigned to the PR.
- [**PR Is Linked to a Milestone**](#pr-is-linked-to-a-milestone): Ensures the PR is associated with a milestone.
- [**PR Changed X Files or Less**](#pr-changed-x-files-or-less): Validates that the number of changed files is within acceptable limits.
- [**PR Has Been Open for X Days**](#pr-has-been-open-for-x-days): Monitors how long a PR has been open.
- [**PR Batch Size Calculation**](#pr-batch-size-calculation): Calculates the batch size of the PR.

These checks are implemented using Port's [scorecards](/#scorecards) on the `pull request` blueprint.

> ## Implementation 

Reviewer: Do you think it might make more sense to add all the metrics in one snipper?
One snipper for scorecard, one for mapping, and one for the properties.

The guide feels complicated where it can easily be a copy-paste one. WDYT?

Author: I think doing that would not leave much of a choice for the user, but to use all these metrics weather they like it or not, it it also doesn't explicitly explain what each json or mapping config is doing. I think is best if each agreement gets a spot light

Reviewer: OK

    "Small": "lightGray",
    "Medium": "orange",
    "Large": "yellow",
    "Gigantic": "red"
  }
}
```
</details>

<h4> Mapping Configuration </h4>
Map the PR's `additions`, `deletions`, and `changedFiles` properties to the `batchSize` property:

<details>
  <summary>Click to view the mapping configuration</summary>

```yaml showLineNumbers
    - kind: pull-request
      selector:
        query: 'true'
        port:
        entity:
            mappings:
            identifier: .head.repo.name + (.id|tostring)
            title: .title
            blueprint: '"githubPullRequest"'
            properties:
                #other properties
            batchSize: >
             if (.commits <= 3 and (.additions + .deletions) <= 50 and
             .changed_files <= 3) then "Small" elif (.commits <= 10 and
             (.additions + .deletions) <= 200 and .changed_files <= 7) then
             "Medium" elif (.commits <= 20 and (.additions + .deletions) <= 500
             and .changed_files <= 15) then "Large" else "Gigantic" end
```
</details>

:::note Batch size calculation
You can adjust the thresholds based on your team's requirements.
:::

## Pull request metrics aggregation

> To measure PR standards effectively, add aggregation properties on the **service** and **organization** blueprints.

Reviewer: ```suggestion
To measure PR standards effectively, add aggregation properties on the **service**  blueprint.
```

You mention organization later, and it's not clear the user even has organization blueprint so I think we can remove it from this sentence

<details>
  <summary>Click to view the calculation property</summary>

```json
"totalLocChanged": {
  "title": "Total LOC Changed",
  "type": "number",
  "calculation": ".properties.additions + .properties.deletions"
}
```

</details>

Add the following aggregation property to the **service** and **organization** blueprints:

<details>
  <summary>Click to view the aggregation property</summary>

```json
"averagePrLinesOfCode": {
  "title": "Average PR Lines of Code",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "func": "average",
    "averageOf": "week",
    "calculationBy": "property",
    "property": "totalLocChanged"
  }
}
```

</details>

</TabItem>

</Tabs>

By implementing these aggregation properties, you can effectively measure and monitor PR standards at both the service and organization levels.

## Aggregation on team level

To measure PR standards at the team level, add the following aggregation properties to the **team** [blueprint](/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples#map-repositories-and-teams):

<Tabs

>   defaultValue="totalAverageCommitsPerPr"

Reviewer: What does "Total average" means? isn't it "average"?

Author: Yeah I think total being there should be removed

    }
  }

```
</details>

</TabItem>

<TabItem value="monthlyAveragePrLoc">

Add the following aggregation property to calculate the monthly average PR lines of code:

<details>
  <summary>Click to view the aggregation property</summary>

```json showLineNumbers
  "monthly_average_pr_loc": {
    "title": "Monthly Average PR LOC",
    "type": "number",
    "target": "service",
    "calculationSpec": {
      "func": "average",
      "averageOf": "month",
      "property": "averagePrLinesOfCode",
      "measureTimeBy": "$createdAt",
      "calculationBy": "property"
    }
  }

```
</details>

</TabItem>

<TabItem value="averageTimeToMerge">

Add the following aggregation property to calculate the average time to merged PRs:

<details>
  <summary>Click to view the aggregation property</summary>

```json showLineNumbers
"average_time_to_merge": {
  "title": "Average Time to Merge",
  "icon": "DefaultProperty",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "func": "average",
    "averageOf": "day",
    "property": "days_old",
    "measureTimeBy": "mergedAt",
    "calculationBy": "property"
  }
}

```
</details>

</TabItem>

</Tabs>

## Visualization

> By leveraging Port's Dashboards, you can create custom dashboards to track the pr metrics and monitor your team's performance over time.

Reviewer: You can also add a simple visualization that group the scorecards by team/service and show the scored (without special aggregation, just a type of widget)

Author: I don't understand this, if you could elaborate on this a bit more

Reviewer: Assuming you have a scorecard score for each service, and each service has on owning team, you can create a table view of all services groups by team and showing average of the teams services scorecard. It's a simple widget but powerful

Reviewer: @kodjomiles ?

Author: okay I will add this so we can merge

### Dashboard setup

> <img src="/img/guides/prMetricsDashboardComp.png" />

Reviewer: This image is nice, but ideally, you can add 1-2 more images showing how it looks like. specifically the widget that shows group by service/team (my comment above), and maybe something like average lines of codes

Author: like this ? 
<img width="1565" alt="image" src="https://github.com/user-attachments/assets/893115ae-a5cc-45b8-965f-a8bd53fed457">

Reviewer: Perfect

---
sidebar_position: 1
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

> # Implement working agreements and measure pull request standards 

Reviewer: I'm missing meaningful screenshots showing both how a PR scorecard looks like, and how an example dashboard looks like (you can take the "scorecard overview" from the demo for example)

## Overview

Effective collaboration and clear expectations are crucial for high-performing engineering teams. **Working agreements** establish shared processes and standards, enhancing teamwork.
Measuring **Pull request (PR)** standards is essential for assessing code quality, review processes, and team efficiency. 
By integrating working agreements with measurable PR metrics, teams can monitor adherence to best practices and continuously improve workflows. 
We will discuss how to implement working agreements and measure PR standards using Port.

:::info Metrics
Metrics are essential for assessing how well teams adhere to their working agreements. 
They enable teams to track compliance, identify bottlenecks, and drive continuous improvement. 
For detailed insights into key metrics like `deployment frequency`, `lead time for changes`, and `change failure rate`,
please refer to our [DORA Metrics guide](/guides/all/setup-dora-metrics).
:::

## Prerequisites
- Complete the [Port onboarding process](https://docs.getport.io/quickstart).
- Access to a GitHub repository that is connected to Port via the onboarding process.

## Working agreements

> The following working agreements and PR checks have been implemented in our [demo environment](https://demo.getport.io/settings/data-model) on the `pull request` blueprint:

Reviewer: maybe the link can lead to a more meaningful page? like the "socrecards overview" page"? or a specific pull request scorecards?

Author: I don't see any scorecard overview page can you point me to the right page

Reviewer: You're right it disappeared. 
Look at https://demo.getport.io/engineering_overview all the way down to the bottom for a similar example.

Author: @Matanga1-2 so we write `The following working agreements and PR checks have been implemented in our [demo environment](https://demo.getport.io/engineering_overview) all the way down for similar examples:` ??

Author: @Matanga1-2  will this be okay then `https://demo.getport.io/my_teams_scorecards`

Author: https://demo.getport.io/scorecard_overview is active now

- [**PR Description Cannot be Empty**](#pr-description-cannot-be-empty): Ensures that every PR has a description.
- [**PR Has Linked Issue**](#pr-has-linked-issue): Verifies that each PR is linked to an issue.
- [**PR Has No Unchecked Checkboxes**](#pr-has-no-unchecked-checkboxes): Checks that there are no unchecked items in the PR description.
- [**PR Requires Reviewers**](#pr-requires-reviewers): Confirms that at least one reviewer is assigned to the PR.
- [**PR Is Linked to a Milestone**](#pr-is-linked-to-a-milestone): Ensures the PR is associated with a milestone.
- [**PR Changed X Files or Less**](#pr-changed-x-files-or-less): Validates that the number of changed files is within acceptable limits.
- [**PR Has Been Open for X Days**](#pr-has-been-open-for-x-days): Monitors how long a PR has been open.
- [**PR Batch Size Calculation**](#pr-batch-size-calculation): Calculates the batch size of the PR.

These checks are implemented using Port's [scorecards](/#scorecards) on the `pull request` blueprint.

## Implementation 
Here are the implementation details for each working agreement and PR check:

### PR Description Cannot be Empty

> <h4> Scorecard Definition </h4>

Reviewer: @kodjomiles you explain here each scorecard, but its' unclear what to do with it.
MAybe you should add a section explaining that you can go to your pull request blueprint, click on add scorecard, and paste each JSON.
Also share screenshot

             and .changed_files <= 15) then "Large" else "Gigantic" end
```
</details>

:::note Batch size calculation
You can adjust the thresholds based on your team's requirements.
:::

## Pull request metrics aggregation

To measure PR standards effectively, add aggregation properties on the **service**  blueprint.
This will allow us to capture important metrics such as:

- PR Average Duration (Service Level)
- PRs Opened (Service & Organization Level)
- PRs Merged (Service & Organization Level)
- Average Commits per PR (Service & Organization Level)
- Average Lines of Code (LOC) Changed (Service & Organization Level)

::::info Aggregation on organization level
To aggregate on an organization level, apply the same settings to a higher hierarchy.
::::

<Tabs
  defaultValue="averagePrDuration"
  values={[
    {label: 'Average PR Duration', value: 'averagePrDuration'},
    {label: 'PRs Opened', value: 'prsOpened'},
    {label: 'PRs Merged', value: 'prsMerged'},
    {label: 'Average Commits per PR', value: 'averageCommitsPerPr'},
    {label: 'Average LOC Changed', value: 'averageLocChanged'},
  ]}>

<TabItem value="averagePrDuration">

> Add the following aggregation property to the **service** blueprint:

Reviewer: @kodjomiles maybe simila to the DORA metric guide explain how to add aggregation prop?

<details>
  <summary>Click to view the aggregation property</summary>

```json
"averagePrDuration": {
  "title": "Average PR Duration",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "func": "average",
    "averageOf": "week",
    "calculationBy": "property",
    "property": "days_old"
  },
  "query": {
    "combinator": "and",
    "rules": [
      {
        "property": "mergedAt",
        "operator": "isNotEmpty"
      }
    ]
  }
}
```

</details>

</TabItem>

<TabItem value="prsOpened">

> Add the following aggregation property to the **service** and **organization** blueprint:

Reviewer: why mentioning "organization" here? we didn't instruct to have or add such blueprint

Author: it was in the notion documentation to aggregate on service and organization

Reviewer: Yes but earlier you mentioned that the aggregations are on service and can be added to higehr hierarchies as well (which is great)

Here it's out of context, the user don't know what is this blueprint

Author: removed

        "operator": "=",
        "value": "open"
      }
    ]
  }
}
```

</details>

</TabItem>

<TabItem value="prsMerged">

Add the following aggregation property to the **service** blueprint:

<details>
  <summary>Click to view the aggregation property</summary>

```json
"mergedPrs": {
  "title": "Merged PRs",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "func": "count",
    "calculationBy": "entities"
  },
  "query": {
    "combinator": "and",
    "rules": [
      {
        "property": "status",
        "operator": "=",
        "value": "merged"
      }
    ]
  }
}
```

</details>

</TabItem>

<TabItem value="averageCommitsPerPr">

Add the following aggregation property to the **service** blueprint:

<details>
  <summary>Click to view the aggregation property</summary>

```json
"averageCommitsPerPr": {
  "title": "Average Commits per PR",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "averageOf": "week",
    "func": "average",
    "calculationBy": "property",
    "property": "commits",
    "measureTimeBy": "$createdAt"
  }
}
```

</details>

</TabItem>

<TabItem value="averageLocChanged">

> First, ensure you have a calculation property `totalLocChanged` on the **pull request** blueprint:

Reviewer: @kodjomiles why not moving this step above to where you define the scorecard? less context switching

Author: okay sounds good

Add the following aggregation property to the **service** and **organization** blueprints:

<details>
  <summary>Click to view the aggregation property</summary>

```json
"averagePrLinesOfCode": {
  "title": "Average PR Lines of Code",
  "type": "number",
  "target": "githubPullRequest",
  "calculationSpec": {
    "func": "average",
    "averageOf": "week",
    "calculationBy": "property",
    "property": "totalLocChanged"
  }
}
```

</details>

</TabItem>

</Tabs>

By implementing these aggregation properties, you can effectively measure and monitor PR standards at both the service and organization levels.

## Aggregation on team level

To measure PR standards at the team level, add the following aggregation properties to the **team** [blueprint](/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples#map-repositories-and-teams):

<Tabs
  defaultValue="averageCommitsPerPr"
  values={[
    {label: 'Monthly Average Commits Per PR', value: 'averageCommitsPerPr'},
    {label: 'Monthly Average PR LOC', value: 'monthlyAveragePrLoc'},
    {label: 'Average Time to Merge', value: 'averageTimeToMerge'}
  ]}>

<TabItem value="averageCommitsPerPr">

> Add the following aggregation property to calculate the total average commits per PR:

Reviewer: @kodjomiles are these any difference from the service ones?

Author: not very significant. should I remove them ?

Reviewer: I would avoid duplication and right above when adding to service under a tip panel that
1. If you want to see it on a team level, you need to make sure services are related to teams and add the same aggregations to team

Author: okay got it

<details>
  <summary>Click to view the aggregation property</summary>

```json showLineNumbers
  "total_average_commits_per_pr": {

>     "title": "Total Average Commits Per PR",

Reviewer: Total average?
